src/RenderWebGL.js

```diff
@@ -653,94 +669,115 @@ class RenderWebGL extends EventEmitter {
-                {idFilterFunc: testID => testID !== drawableID}
-            );
-        } finally {
-            gl.colorMask(true, true, true, true);
-            gl.disable(gl.STENCIL_TEST);
-        }
+                gl.stencilFunc(gl.EQUAL, 1, 1);
+                gl.stencilOp(gl.KEEP, gl.KEEP, gl.KEEP);
+                gl.colorMask(true, true, true, true);
 
-        const pixels = new Uint8Array(Math.floor(bounds.width * bounds.height * 4));
-        gl.readPixels(0, 0, bounds.width, bounds.height, gl.RGBA, gl.UNSIGNED_BYTE, pixels);
+                this._drawThese(candidateIDs, ShaderManager.DRAW_MODE.default, projection,
+                    {idFilterFunc: testID => testID !== drawableID}
+                );
+            } finally {
+                gl.colorMask(true, true, true, true);
+                gl.disable(gl.STENCIL_TEST);
+            }
 
-        if (this._debugCanvas) {
-            this._debugCanvas.width = bounds.width;
-            this._debugCanvas.height = bounds.height;
-            const context = this._debugCanvas.getContext('2d');
-            const imageData = context.getImageData(0, 0, bounds.width, bounds.height);
-            imageData.data.set(pixels);
-            context.putImageData(imageData, 0, 0);
-        }
+            const pixels = new Uint8Array(Math.floor(bounds.width * bounds.height * 4));
+            gl.readPixels(0, 0, bounds.width, bounds.height, gl.RGBA, gl.UNSIGNED_BYTE, pixels);
 
-        for (let pixelBase = 0; pixelBase < pixels.length; pixelBase += 4) {
-            const pixelDistanceR = Math.abs(pixels[pixelBase] - color3b[0]);
-            const pixelDistanceG = Math.abs(pixels[pixelBase + 1] - color3b[1]);
-            const pixelDistanceB = Math.abs(pixels[pixelBase + 2] - color3b[2]);
+            if (this._debugCanvas) {
+                this._debugCanvas.width = bounds.width;
+                this._debugCanvas.height = bounds.height;
+                const context = this._debugCanvas.getContext('2d');
+                const imageData = context.getImageData(0, 0, bounds.width, bounds.height);
+                imageData.data.set(pixels);
+                context.putImageData(imageData, 0, 0);
+            }
 
-            if (pixelDistanceR <= TOLERANCE_TOUCHING_COLOR.R &&
-                pixelDistanceG <= TOLERANCE_TOUCHING_COLOR.G &&
-                pixelDistanceB <= TOLERANCE_TOUCHING_COLOR.B) {
-                return true;
+            for (let pixelBase = 0; pixelBase < pixels.length; pixelBase += 4) {
+                if (colorMatches(color3b, pixels, pixelBase)) {
+                    return true;
+                }
             }
+
+            return false;
         }
 
+        const drawable = this._allDrawables[drawableID];
+        const point = __isTouchingDrawablesPoint;
+        const color = __touchingColor;
+        const hasMask = Boolean(mask3b);
+        for (let x = bounds.left; x <= bounds.right; x++) {
```

Comment:
There's a chance that this would be more cache-friendly by swapping the loops so that `x` is the inner loop. I suspect there won't be a measurable difference, especially with a size cap on the CPU-rendered path, but you never know...